server/rest/instance.py

```diff
@@ -222,3 +231,55 @@ def handleUpdateJob(self, event):
         elif status in (JobStatus.QUEUED, JobStatus.RUNNING):
             instance['status'] = InstanceStatus.LAUNCHING
         self._model.updateInstance(instance)
+
+    @access.public
+    @autoDescribeRoute(
+        Description('Initiate oauth login flow')
+        .param('redirect', 'URL to redirect to after login', required=False)
+    )
+    def authenticate(self, redirect):
+        # TODO: Probably doesn't belong on instance
+        user = self.getCurrentUser()
+
+        # If there's no user, initiate the oauth flow with this endpoint
+        # as the callback along with the fhost parameter
+        if user is None:
+            oauth_providers = OAuthResource().listProviders(params={"redirect": redirect})
+            raise cherrypy.HTTPRedirect(oauth_providers["Globus"])
+        else:
+            raise cherrypy.HTTPRedirect(redirect)
+
+    @access.cookie
+    @access.public
+    @autoDescribeRoute(
+        Description('Determine whether user has access to instance requested via forward auth')
+    )
+    def authorize(self):
+        # This endpoint must be called frmo a Traefik forward-auth request. The X-Forwarded-Host
+        # is assumed to be the hostname for a running instance. Also assumes that the
+        # core.cookie_domain is set to .(local.)wholetale.org
+        user = self.getCurrentUser()
+
+        forwarded_host = cherrypy.request.headers.get('X-Forwarded-Host')
+        forwarded_uri = cherrypy.request.headers.get('X-Forwarded-Uri')
+        if not forwarded_host and not forwarded_uri:
+            raise RestException('Forward auth request required', code=400)
+
+        if user is None:
+            # If no user, redirect to authentication endpoint to initiate oauth flow
+            domain = forwarded_host.partition('.')[2]
+            redirect = 'https://' + forwarded_host + forwarded_uri 
+            # As a forward-auth request, the host is the origin (e.g., tmp-xxx.*)
+            # but we need to redirect to Girder.
+            raise cherrypy.HTTPRedirect('https://girder.' + domain + '/api/v1/instance/authenticate?redirect=' + redirect)
+
+        instances = list(self._model.list(user=user, currentUser=user))
+        access = False
+        for instance in instances:
+            url = urlparse(instance['url'])
+            if forwarded_host == url.netloc:
+                access = True
+                break
+
+        if not access:
+            raise RestException('Access denied for instance', code=403)
```

Comment:
How about we simply check if there's an instance that matches subdomain and was created by the user?
```suggestion
        forwarded_host = cherrypy.request.headers.get('X-Forwarded-Host')
        forwarded_uri = cherrypy.request.headers.get('X-Forwarded-Uri')
        subdomain, domain = forwarded_host.split(".", 1)
        if not forwarded_host and not forwarded_uri:
            raise RestException('Forward auth request required', code=400)

        if user is None:
            # If no user, redirect to authentication endpoint to initiate oauth flow
            redirect = 'https://' + forwarded_host + forwarded_uri 
            # As a forward-auth request, the host is the origin (e.g., tmp-xxx.*)
            # but we need to redirect to Girder.
            raise cherrypy.HTTPRedirect('https://girder.' + domain + '/api/v1/instance/authenticate?redirect=' + redirect)
 
        if self._model.findOne(
            {"containerInfo.name": subdomain, "creatorId": user["_id"]]}
        ):
            return
        raise RestException('Access denied for instance', code=403)
```

Reply:
Your suggestions look great. Thanks.